verida/vault-mobile
# Feature 977 - Add blockchain mainnets and support for custom blockchains

Tracks the epic outlined in #977.

### 💭 What's changed?

- Added an in-memory abstraction over a `BlockchainAccount` called a `MinifiedVeridaAccount`.
  - To avoid the effort of migrating the persistent datastore away from the private key duplication architecture, we instead introduce the concept of a `MinifiedVeridaAccount`. This represents the minimum amount of information required to populate a `Signer` for each `namespace`, and in this representation, we are easily able to filter duplicates.
  - Using an in-memory implementation affords us time to play with the abstraction and see if it works for us before committing it to the persistence layer.
  - In general, anywhere we are using a `BlockchainAccount` in the Vault App, we can now start being skeptical about (at least, all I care about so far in this app is crypto - it may have more important relevance elsewhere). The reason for this is because a `BlockchainAccount` is too tightly coupled to the configuration of a certain chain identifier, and it's app-specific usage may often no longer be semantically correct for the account - many new areas of the app expect to be able to use a certain private key with an arbitrary chain, which a `privateKey` supports but a `BlockchainAccount` technically does not).
  - This provides us with a way of representing unique private keys, regardless of whether they are duplicated in memory per networks, and sufficiently abstracts them away from a specific chain context that they can be _reused_ for chains which aren't explicitly listed inside of persistent storage.
  - In effect, we can take the default Goerli/Polygon/etc private keys, dedup these into unique private keys, then use them across any EIP-155 network.
  - Uses similar implementation logic for NEAR.
- Removed account creation logic for NEAR.
  - We encountered multiple issues associated with the WalletProvider which emphasises the app's unreadiness for Near Accounts (alongside the unintentional lock-in of deterministic AccountID generation). Transactions have been refactored to go through the `signerId` instead, which evaluates consistent NEAR balances against the WalletProvider. Code related to accounts has been removed (but can be referred back to later).
- Extended the implementation of `ChainMetadata` to contain desired properties for network configuration.
  - Currently, `ChainMetadata` is not a `CustomBlockchainNetwork`, so we need a way to translate back-and-forth between them (the app talks at runtime in terms of `ChainMetadata`, but Verida datastores persist `CustomBlockchainNetwork`s instead). We can easily replace `ChainMetadata` with `CustomBlockchainNetwork` in a future refactor.
-  Added network configuration screen under Settings.
- Removed old transaction architecture which used to go through both Redux and the Verida backend.
  - Transfers are now submitted entirely natively, and shares most of the implementation logic with WalletConnect.
- Added ability to add new networks via WalletConnect, programatically on iOS simulator via `yarn wc`, and manually.
  - Also added support for `wallet_addEthereumChain` on EIP155, which is a subset of the generic blockchain storage model.
- Added locale-specific blockchain networks.
  - In `Australia`, users have the `RedBelly Network` available to them. Can be extended arbitrarily for each country.
-  Refactored the `WalletConnectModal` into an `Adapter` pattern.
  - This allows us to programmatically trigger requests within the context of the application and override the internal display logic with custom components. This is useful for automated requests like `wallet_addEthereumChain`.
- Updated `TokenCalculator` so that core business logic could be evaluated and depend upon from multiple parts of the app instead of explicitly through deferred stateful render callbacks. This allows more consistent handling.
- Doubled-down on the concept of a Currency.
  - The values included are exported as a `BigDecimal` to achieve more accurate currency calculations and demands much stricter interactions with related logic.
  - Currently, we only support `USD`. 
- Removed a lot of (eventually) unused code.
- Added the ability to trigger a PaymentRequest on iOS via `yarn pay`.
- Migrated all affected regions over to TypeScript.
- Added placeholder functionality for gas estimation indicators - this needs further work.
  - For now, since the gas indicator is mostly used on transfers, here we just assume this to be the network transfer fee, but the core architecture can be easily extended.

### 🧪 How to test these changes?

- Demands a full regression test of all features relating to wallets, accounts and networks.
- Play around a lot with WalletConnect: https://react-app.walletconnect.com/
  - Connect to new networks, reject requests to save, accept, etc.
- Try to go through the PaymentRequest flow and ensure these are consistent for native coins and ERC20s.
  - @aurelticot, I really hope I managed to capture your intention on these screens! The project was in complete disarray when trying to solve merge conflicts, which led to me needing to rebuild a lot of the flow to use the native transaction functions. But I lack a lot of the required knowledge here, so hopefully you can identify anythings that's missing and even more hopefully it will be easily fixed!
- Add your own custom networks, modify them, delete them.
  - You can now run `anvil` on your local machine and transact with a local blockchain. You can also use `cast` to credit accounts in your wallet.
  - Try to transact on the networks by sending some native currency. Check how the UI responds - especially the visibility of certain characteristics on price. There's a lot of price rendering logic that I had to refactor, so hopefully this is still consistent.
- Really play around with the price control, swapping from fiat to crypto, and ensure the price renders as expected. 

### 😨 Anything to be aware of?

- A lot of big PRs have been merged whilst this was in progress. I may easily have regressed an important recent change, so these are something to watch out for.
- We decided against further bloating this incredibly bloated PR with the Custom Blockchain Network Verida Message functionality and the custom graphics in the Add New Network window.
  - These can be delivered in future.
